contrib/gperl/gperl.man
.TH GPERL @MAN1EXT@ "@MDATE@" "Groff Version @VERSION@"
.SH NAME
gperl \- groff preprocessor for Perl parts in roff files
.
.\" The .SH was moved to this place in order to appease `apropos'.
.
.\" --------------------------------------------------------------------
.\" Legalese
.\" --------------------------------------------------------------------
.
.de au
This file was written by
.MT groff-bernd.warken-72@web.de
Bernd Warken
.ME .
..
.
.de co
Copyright \(co 2014 Free Software Foundation, Inc.
.
.P
This file is part of
.IR \%gperl ,
which is part of
.IR \%groff ,
a free software project.
.
You can redistribute it and/or modify it under the terms of the
.nh
.B "GNU General Public License"
.hy
as published by the
.nh
.BR "Free Software Foundation" ,
.hy
version\~2.
.
.P
The license text is available in the internet at
.UR http://www.gnu.org/licenses/gpl-2.0.html
.UE .
..
.
.\" --------------------------------------------------------------------
.\" Characters
.\" --------------------------------------------------------------------
.
.\" Ellipsis ...
.ie t .ds EL \fS\N'188'\fP\"
.el .ds EL \&.\|.\|.\&\"
.\" called with \*(EL
.
.\" Bullet
.ie t .ds BU \(bu
.el .ds BU *
.\" used in `.IP \*(BU 2m' (former .Topic)
.
.
.\" --------------------------------------------------------------------
.SH "SYNOPSIS"
.\" --------------------------------------------------------------------
.
.SY gperl
.OP \-
.OP \-\-
.OP \& "\%filespec \*(EL"
.YS
.
.BR "gperl -h" | --help
.br
.BR "gperl -v" | --version
.
.
.\" --------------------------------------------------------------------
.SH DESCRIPTION
.\" --------------------------------------------------------------------
.
This is a preprocesor for
.BR \%groff (@MAN1EXT@).
.
It allows to add
.BR perl (7)
code into
.BR groff (7)
files.
.
The result of a
.I Perl part
can be stored in groff
.I strings
or
.I numerical registers
based on the arguments at a final line of a
.IR "Perl part" .
.
.
.\" --------------------------------------------------------------------
.SH "OPTIONS"
.\" --------------------------------------------------------------------
.
So far, there are only
.I filespec
or
.I breaking
options.
.
.
.P
.I filespec
are file names or the minus character
.B \-
character for standard input.
.
As usual, the argument
.B \-\-
can be used in order to let all fowllowing arguments mean file names,
even if the names begin with a minus character
.BR \- .
.
.
.P
An option is
.IR breaking ,
when the program just writes the information that was asked for and
then stops.
.
All other arguments will be ignored by that.
.
These
.I breaking
options are heree
.
.TP
.B -h\~\fR|\fB\~--help
Print help information with a short explanation of options to
standard output.
.
.
.TP
.B -v\~\fR|\fB\~--version
Print version information to standard output.
.
.
.\" --------------------------------------------------------------------
.SH "PERL PARTS"
.\" --------------------------------------------------------------------
.
.I Perl
parts in
.I groff files
are enclosed by two
.B .Perl
requests with different arguments, a
.I starting
and an
.I ending
command.
.
.
.\" --------------------------------------------------------------------
.SS "Starting Perl Mode"
.\" --------------------------------------------------------------------
.
The starting
.I Perl request
can either be without arguments, or by a request that has the term
.B start
as its only argument.
.RS
.IP \*(BU 2m
.B \&.Perl
.IP \*(BU 2m
.B \&.Perl start
.RE
.
.
.\" --------------------------------------------------------------------
.SS "Ending Perl Mode without Storage"
.\" --------------------------------------------------------------------
.
A
.B .Perl
command line with an argument different from
.B start
finishes a running
.IR "Perl part" .
.
Of course, it would be reasonable to add the argument
.BR stop ;
that's possible, but not necessary.
.
.RS
.IP \*(BU 2m
.B \&.Perl stop
.IP \*(BU 2m
.BI \&.Perl " other_than_start"
.RE
.
The argument
.I other_than_start
can additionally be used as a
.I groff
string variable name for storage \(em see next section.
.
.
.\" --------------------------------------------------------------------
.SS "Ending Perl Mode with Storage"
.\" --------------------------------------------------------------------
.
A useful feature of
.B gperl
is to store one or more results from the
.IR "Perl mode" .
.
.
.P
The output of a
.I Perl part
can be got with backticks
.BR `...` .
.
.
.P
This program collects all printing to STDOUT (normal standard output)
by the Perl
.B print
program.
.
This pseudo-printing output can have several lines, due to printed
line breaks with
.BR \(rsn .
.
By that, the output of a Perl run should be stored into a Perl array,
with a single line for each array member.
.
.
.P
This Perl array output can be stored by
.B gperl
in either
.TP
.I groff strings
by creating a groff command
.B .ds
.
.TP
.I groff number register
by creating a groff command
.B .rn
.
.
.P
The storage modes can be determined by arguments of a final stopping
.B .Perl
command.
.
Each argument
.B .ds
changes the mode into
.I groff string
and
.B .nr
changes the mode into
.I groff number register
for all following output parts.
.
.
.P
By default, all output is saved as strings, so
.B .ds
is not really needed before the first
.B .nr
command.
.
That suits to
.BR \%groff (@MAN7EXT@),
because every output can be saved as
.I groff
string, but the number registers can be very restrictive.
.
.
.P
In
.IR "string mode" ,
.B gperl
generates a
.I groff string
storage line
.RS
.EX
\&\fB.ds \fIvar_name content
.EE
.RE
.
In
.I number register mode
the following groff command is generated
.RS
.EX
\&\fB.nr \fIvar_name content
.EE
.RE
.
.
.P
We present argument collections in the following.
.
You can add as first argument for all
.BR stop .
.
We omit this additional element. 
.
.
.P
.TP
.BI ".Perl .ds " var_name
This will store 1 output line into the groff string named
.I var_name
by the automatically created command
.RS
.RS
.EX
.BI .ds " var_name output"
.EE
.RE
.RE
.
.
.TP
.BI .Perl " var_name"
If
.I var_name
is different from
.B start
this is equivalent to the former command, because the string mode is
string with
.B .ds
command.
default.
.
.
.TP
.BI .Perl " var_name1 var_name2"
This will store 2 output lines into groff string names
.I var_name1
and
.IR var_name2 ,
because the default mode
.B .ds
is active, such that no
.B .ds
argument is needed.
.
Of course, this is equivalent to
.RS
.RS
.EX
.BI ".Perl .ds " "var_name1 var_name2"
.EE
.RE
and
.RS
.EX
.BI ".Perl .ds " "var_name1 " ".ds" " var_name2"
.EE
.RE
.RE
.
.
.TP
.BI ".Perl .nr" " var_name1 varname2"
stores both variables as number register variables.
.
.B gperl
generates
.RS
.EX
.BI .nr " var_name1 output_line1"
.BI .nr " var_name2 output_line2"
.EE
.RE
.
.
.TP
.BI ".Perl .nr " var_name1 " .ds " var_name2
stores the 1st argument as
.I number register
and the second as
.I string
by
.RS
.EX
.BI .nr " var_name1 output_line1"
.BI .ds " var_name2 output_line2"
.EE
.RE
.
.
.\" --------------------------------------------------------------------
.SS "Printing towards STDERR is without Storage"
.\" --------------------------------------------------------------------
.
The printing towards
.IR STDERR ,
(standard error) works as usual.
.
All error information goes to the real normal
.IR "standard error" ,
without other automatical storage.
.
.
.\" --------------------------------------------------------------------
.SH "EXAMPLES"
.\" --------------------------------------------------------------------
.
A possible
.I Perl part
in a
.I roff file
could look like that:
.RS
.EX
before
\&.Perl start
my $result = 'some data';
print $result;
\&.Perl stop .ds string_var
after
.EE
.RE
.
.
.P
This stores the result
.B \(rqsome data\(rq
into the
.I roff string
called
.BR string_var ,
such that the following line is printed:
.RS
.EX
\&.ds string_var some data
.EE
.RE
by
.B gperl
as food for the coming
.B groff
run.
.
.
.P
A
.I Perl part
with several outputs is:
.RS
.EX
\&.Perl start
print \(rqfirst\(rsn\(rq;
print \(rqsecond line\(rsn\(rq;
print \(rq3\(rsn\(rq;
\&.Perl var1 var2 .nr var3 
.EE
.RE
.
This stores 3 printed lines into 3
.I groff
strings.
.BR var1 , var2 , var3 .
.
So the following
.I groff
command lines are created:
.RS
.EX
\&.ds var1 first
\&.ds var2 second line
\&.nr var3 3
.EE
.RE
.
.
.\" --------------------------------------------------------------------
.SH "SEE ALSO"
.\" --------------------------------------------------------------------
.
.P
Man\-pages related to
.I groff
are
.BR \%groff (@MAN1EXT@),
.BR \%groff (@MAN7EXT@),
.BR \%grog (@MAN1EXT@),
and
.BR \%groffer (@MAN1EXT@).
.
.
.P
Documents related to
.I Perl
are
.BR \%perl (@MAN1EXT@),
.BR \%perl (@MAN7EXT@).
.
.
.\" --------------------------------------------------------------------
.SH "COPYING"
.\" --------------------------------------------------------------------
.co
.\" --------------------------------------------------------------------
.SH "AUTHORS"
.\" --------------------------------------------------------------------
.au
.
.
.\" --------------------------------------------------------------------
.\" Emacs settings
.\" --------------------------------------------------------------------
.
.\" Local Variables:
.\" mode: nroff
.\" End:


contrib/gperl/gperl.pl
#! /usr/bin/env perl

# gperl - add Perl part to groff files, this is the preprocessor for that

# Source file position: <groff-source>/contrib/gperl/gperl.pl
# Installed position: <prefix>/bin/gperl

# Copyright (C) 2014  Free Software Foundation, Inc.

# Written by Bernd Warken <groff-bernd.warken-72@web.de>.

my $version = '1.2.6';

# This file is part of `gperl', which is part of `groff'.

# `groff' is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.

# `groff' is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.

# You can find a copy of the GNU General Public License in the internet
# at <http://www.gnu.org/licenses/gpl-2.0.html>.

########################################################################

use strict;
use warnings;
#use diagnostics;

# temporary dir and files
use File::Temp qw/ tempfile tempdir /;

# needed for temporary dir
use File::Spec;

# for `copy' and `move'
use File::Copy;

# for fileparse, dirname and basename
use File::Basename;

# current working directory
use Cwd;

# $Bin is the directory where this script is located
use FindBin;


########################################################################
# system variables and exported variables
########################################################################

$\ = "\n";	# final part for print command

########################################################################
# read-only variables with double-@ construct
########################################################################

our $File_split_env_sh;
our $File_version_sh;
our $Groff_Version;

my $before_make;		# script before run of `make'
{
  my $at = '@';
  $before_make = 1 if '@VERSION@' eq "${at}VERSION${at}";
}

my %at_at;
my $file_perl_test_pl;
my $groffer_libdir;

if ($before_make) {
  my $gperl_source_dir = $FindBin::Bin;
  $at_at{'BINDIR'} = $gperl_source_dir;
  $at_at{'G'} = '';
} else {
  $at_at{'BINDIR'} = '@BINDIR@';
  $at_at{'G'} = '@g@';
}


########################################################################
# options
########################################################################

foreach (@ARGV) {
  if ( /^(-h|--h|--he|--hel|--help)$/ ) {
    print q(Usage for the `gperl' program:);
    print 'gperl [-] [--] [filespec...] normal file name arguments';
    print 'gperl [-h|--help]        gives usage information';
    print 'gperl [-v|--version]     displays the version number';
    print q(This program is a `groff' preprocessor that handles Perl ) .
      q(parts in `roff' files.);
    exit;
  } elsif ( /^(-v|--v|--ve|--ver|--vers|--versi|--versio|--version)$/ ) {
    print q(`gperl' version ) . $version;
    exit;
  }
}


#######################################################################
# temporary file
#######################################################################

my $out_file;
{
  my $template = 'gperl_' . "$$" . '_XXXX';
  my $tmpdir;
  foreach ($ENV{'GROFF_TMPDIR'}, $ENV{'TMPDIR'}, $ENV{'TMP'}, $ENV{'TEMP'},
	   $ENV{'TEMPDIR'}, 'tmp', $ENV{'HOME'},
	   File::Spec->catfile($ENV{'HOME'}, 'tmp')) {
    if ($_ && -d $_ && -w $_) {
      eval { $tmpdir = tempdir( $template,
				CLEANUP => 1, DIR => "$_" ); };
      last if $tmpdir;
    }
  }
  $out_file = File::Spec->catfile($tmpdir, $template);
}


########################################################################
# input
########################################################################

my $perl_mode = 0;

foreach (<>) {
  chomp;
  s/\s+$//;
  my $line = $_;
  my $is_dot_Perl = $line =~ /^[.']\s*Perl(|\s+.*)$/;

  unless ( $is_dot_Perl ) {	# not a `.Perl' line
    if ( $perl_mode ) {		# is running in Perl mode
      print OUT $line;
    } else {			# normal line, not Perl-related
      print $line;
    }
    next;
  }


  ##########
  # now the line is a `.Perl' line

  my $args = $line;
  $args =~ s/\s+$//;	# remove final spaces
  $args =~ s/^[.']\s*Perl\s*//;	# omit .Perl part, leave the arguments

  my @args = split /\s+/, $args;

  ##########
  # start Perl mode
  if ( @args == 0 || @args == 1 && $args[0] eq 'start' ) {
    # For `.Perl' no args or first arg `start' means opening `Perl' mode.
    # Everything else means an ending command.
    if ( $perl_mode ) {
      # `.Perl' was started twice, ignore
      print STDERR q(`.Perl' starter was run several times);
      next;
    } else {	# new Perl start
      $perl_mode = 1;
      open OUT, '>', $out_file;
      next;
    }
  }

  ##########
  # now the line must be a Perl ending line (stop)

  unless ( $perl_mode ) {
    print STDERR 'gperl: there was a Perl ending without being in ' .
      'Perl mode:';
    print STDERR '    ' . $line;
    next;
  }

  $perl_mode = 0;	# `Perl' stop calling is correct
  close OUT;		# close the storing of `Perl' commands

  ##########
  # run this `Perl' part, later on about storage of the result
  # array stores prints with \n
  my @print_res = `perl $out_file`;

  # remove `stop' arg if exists
  shift @args if ( $args[0] eq 'stop' );

  if ( @args == 0 ) {
    # no args for saving, so @print_res doesn't matter
    next;
  }

  my @var_names = ();
  my @mode_names = ();

  my $mode = '.ds';
  for ( @args ) {
    if ( /^\.?ds$/ ) {
      $mode = '.ds';
      next;
    }
    if ( /^\.?nr$/ ) {
      $mode = '.nr';
      next;
    }
    push @mode_names, $mode;
    push @var_names, $_;
  }

  my $n_res = @print_res;
  my $n_vars = @var_names;

  if ( $n_vars < $n_res ) {
    print STDERR 'gperl: not enough variables for Perl part: ' .
      $n_vars . ' variables for ' . $n_res . ' output lines.';
  } elsif ( $n_vars > $n_res ) {
    print STDERR 'gperl: too many variablenames for Perl part: ' .
      $n_vars . ' variables for ' . $n_res . ' output lines.';
  }
  if ( $n_vars < $n_res ) {
    print STDERR 'gperl: not enough variables for Perl part: ' .
      $n_vars . ' variables for ' . $n_res . ' output lines.';
  }

  my $n_min = $n_res;
  $n_min = $n_vars if ( $n_vars < $n_res );
  exit unless ( $n_min );
  $n_min -= 1; # for starting with 0

  for my $i ( 0..$n_min ) {
    my $value = $print_res[$i];
    chomp $value;
    print $mode_names[$i] . ' ' . $var_names[$i] . ' ' . $value;
  }
}


1;
########################################################################
### Emacs settings
# Local Variables:
# mode: CPerl
# End:
